app/src/main/java/org/thoughtcrime/securesms/groups/GroupsV2Authorization.java

```diff
@@ -0,0 +1,92 @@
+package org.thoughtcrime.securesms.groups;
+
+import androidx.annotation.NonNull;
+
+import org.signal.core.util.logging.Log;
+import org.signal.zkgroup.VerificationFailedException;
+import org.signal.zkgroup.auth.AuthCredentialResponse;
+import org.signal.zkgroup.groups.GroupSecretParams;
+import org.whispersystems.signalservice.api.groupsv2.GroupsV2Api;
+import org.whispersystems.signalservice.api.groupsv2.GroupsV2AuthorizationString;
+import org.whispersystems.signalservice.api.groupsv2.NoCredentialForRedemptionTimeException;
+import org.whispersystems.signalservice.api.push.ACI;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+public final class GroupsV2Authorization {
+
+  private static final String TAG = Log.tag(GroupsV2Authorization.class);
+
+  private final ValueCache  cache;
+  private final GroupsV2Api groupsV2Api;
+
+  public GroupsV2Authorization(@NonNull GroupsV2Api groupsV2Api, @NonNull ValueCache cache) {
+    this.groupsV2Api = groupsV2Api;
+    this.cache       = cache;
+  }
+
+  public GroupsV2AuthorizationString getAuthorizationForToday(@NonNull ACI self,
+                                                              @NonNull GroupSecretParams groupSecretParams)
+      throws IOException, VerificationFailedException
+  {
+    final int today = currentTimeDays();
+
+    Map<Integer, AuthCredentialResponse> credentials = cache.read();
+
+    try {
+      return getAuthorization(self, groupSecretParams, credentials, today);
+    } catch (NoCredentialForRedemptionTimeException e) {
+      Log.i(TAG, "Auth out of date, will update auth and try again");
+      cache.clear();
+    } catch (VerificationFailedException e) {
+      Log.w(TAG, "Verification failed, will update auth and try again", e);
+      cache.clear();
+    }
+
+    Log.i(TAG, "Getting new auth credential responses");
+    credentials = groupsV2Api.getCredentials(today);
+    cache.write(credentials);
+
+    try {
+      return getAuthorization(self, groupSecretParams, credentials, today);
+    } catch (NoCredentialForRedemptionTimeException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)